log-manager/src/main/java/io/airlift/log/BufferedHandler.java

```diff
@@ -1,219 +1,399 @@
-            reportError(null, e, WRITE_FAILURE);
+            // Wrap exception with the proper classification
+            throw new LogFormatException(e);
         }
+    }
 
-        // if closed while queueing, try to remove the message just to clean up
-        if (closed.get()) {
-            queue.remove(message);
+    private static class LogFormatException
+            extends RuntimeException
+    {
+        public LogFormatException(Throwable cause)
+        {
+            super(cause);
         }
     }
 
     @Override
-    public synchronized void flush()
+    public void publish(LogRecord record)
     {
         try {
-            messageOutput.flush();
+            if (!isLoggable(record)) {
+                return;
+            }
+
+            if (inputClosed.get()) {
+                return;
+            }
+
+            // Generate the message on the publishing thread to ensure we get the correct thread name.
+            MessageAndSource message = toMessageAndSource(record);
+
+            // Messages may be inserted after being closed, but they won't be processed if they come after the terminal message
+            queueInsert(message);
+        }
+        catch (LogFormatException e) {
+            reportError(null, e, FORMAT_FAILURE);
         }
         catch (Exception e) {
-            reportError(null, e, FLUSH_FAILURE);
+            reportError(null, e, GENERIC_FAILURE);
         }
     }
 
-    @Override
-    public void close()
+    private MessageAndSource toMessageAndSource(LogRecord record)
     {
-        closed.set(true);
-
-        putUninterruptibly(queue, POISON_MESSAGE);
+        return new MessageAndSource(formatMessageBytes(record), determineSourceName(record));
+    }
 
-        // wait for logging to finish
+    private byte[] formatMessageBytes(LogRecord logRecord)
+    {
         try {
-            thread.join();
+            return getFormatter().format(logRecord).getBytes(UTF_8);
         }
-        catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
+        catch (Exception e) {
+            // Wrap exception with the proper classification
+            throw new LogFormatException(e);
         }
-
-        queue.clear();
     }
 
-    @VisibleForTesting
-    MessageOutput getMessageOutput()
+    private static String determineSourceName(LogRecord record)
     {
-        return messageOutput;
+        return firstNonNull(record.getLoggerName(), "UNKNOWN");
     }
 
-    protected void start()
+    private void queueInsert(MessageAndSource message)
     {
-        thread.start();
+        while (!queue.offerLast(message)) {
+            synchronized (queueDrainLock) {
+                MessageAndSource dropCandidate = queue.peekFirst();
+                if (dropCandidate == null) {
+                    continue;
+                }
+                if (dropCandidate == TERMINAL_MESSAGE) {
+                    // Queue already terminated
+                    return;
```

Comment:
shouldn't this increment `droppedMessages` like the old code did?  If it too annoying, to count messages after close perfectly, just add a comment to the field

Reply:
Based on our discussion, it sounds like we won't be counting the dropped messages after close, since this is typically not what the caller cares about if they are closing the stream explicitly already.